runtime/lua/vim/semver.lua

```diff
@@ -0,0 +1,248 @@
+local M = {}
+
+---@private
+local function cmp_prerelease(v1_parsed, v2_parsed)
+  if v1_parsed.prerelease and not v2_parsed.prerelease then
+    return -1
+  end
+  if not v1_parsed.prerelease and v2_parsed.prerelease then
+    return 1
+  end
+  if not v1_parsed.prerelease and not v2_parsed.prerelease then
+    return 0
+  end
+
+  local v1_identifiers = vim.split(v1_parsed.prerelease, '%.')
+  local v2_identifiers = vim.split(v2_parsed.prerelease, '%.')
+  local i = 1
+  local max = math.max(vim.tbl_count(v1_identifiers), vim.tbl_count(v2_identifiers))
+  while i <= max do
+    local v1_identifier = v1_identifiers[i]
+    local v2_identifier = v2_identifiers[i]
+    if v1_identifier ~= v2_identifier then
+      local v1_num = tonumber(v1_identifier)
+      local v2_num = tonumber(v2_identifier)
+      local is_number = v1_num and v2_num
+      if is_number then
+        -- Number comparisons
+        if not v1_num and v2_num then
+          return -1
+        end
+        if v1_num and not v2_num then
+          return 1
+        end
+        if v1_num == v2_num then
+          return 0
+        end
+        if v1_num > v2_num then
+          return 1
+        end
+        if v1_num < v2_num then
+          return -1
+        end
+      else
+        -- String comparisons
+        if v1_identifier and not v2_identifier then
+          return 1
+        end
+        if not v1_identifier and v2_identifier then
+          return -1
+        end
+        if v1_identifier < v2_identifier then
+          return -1
+        end
+        if v1_identifier > v2_identifier then
+          return 1
+        end
+        if v1_identifier == v2_identifier then
+          return 0
+        end
+      end
+    end
+    i = i + 1
+  end
+
+  return 0
+end
+
+---@private
+local function cmp_version_core(v1_parsed, v2_parsed)
+  if
+    v1_parsed.major == v2_parsed.major
+    and v1_parsed.minor == v2_parsed.minor
+    and v1_parsed.patch == v2_parsed.patch
+  then
+    return 0
+  end
+
+  if
+    v1_parsed.major > v2_parsed.major
+    or v1_parsed.minor > v2_parsed.minor
+    or v1_parsed.patch > v2_parsed.patch
+  then
+    return 1
+  end
+
+  return -1
+end
+
+--- Compares two strings (`v1` and `v2`) in semver format.
+--- @param v1 string version.
+--- @param v2 string version to be compared with v1.
+--- @param opts table|nil Options
+---   Table containing options to configure parsing behaviour. The options and
+---   their default values are shown below:
+---
+---   `{ strict = false (boolean) }`
+---
+---   - `strict`: when set to `true` an error will be thrown for version
+---   strings which are not conforming to semver sepecifications, this means
+---   that `semver.parse('v1.2)` *will throw an error*. When set to `false`
+---   `semver.parse('v1.2)` will coerce 'v1.2' to 'v1.2.0' and return the
+---   table: `{ major = 1, minor = 2, patch = 0 }`.
+--- @returns result number --- -1 if `v1 < v2`,0 if `v1 == v2` or, 1 if `v1 > v2`.
```

Comment:
```suggestion
--- @return result number --- -1 if `v1 < v2`,0 if `v1 == v2` or, 1 if `v1 > v2`.
```

Reply:
`@returns` (plural) is correct syntax, lua doc strings in other parts of the code base is also using `@returns` and the linter throws an error for `@return`.

Reply:
I'm not sure this works with lua_lsp

Reply:
I am using `sumneko_lua` and it is throwing a linting error when I try to use `@return` instead of `@returns`.

I did a quick `grep` through the code base and it looks like `@return` is only every used in `*.c, *.h` files. But all `*.lua` files use `@returns` and `@return` does not appear in any `*.lua` files.

Reply:
Yes, this doesn't work with lua_lsp. You also need to drop `results` and change `number` to `integer`.

Reply:
> I did a quick grep through the code base and it looks like @return is only ever used in *.c, *.h files. But all *.lua files use @returns and @return does not appear in any *.lua files.

Look in `treesitter.lua`.